framework/src/meshgenerators/ParsedExtraElementIDGenerator.C

```diff
@@ -0,0 +1,132 @@
+                                              "subdomains (empty means the entire domain)");
+  params.addParam<std::vector<std::string>>(
+      "constant_names", {}, "Vector of constants used in the parsed function");
+  params.addParam<std::vector<std::string>>(
+      "constant_expressions",
+      {},
+      "Vector of values for the constants in constant_names (can be an FParser expression)");
+  params.addParam<std::vector<ExtraElementIDName>>(
+      "extra_element_id_names", {}, "Extra element integers used in the parsed expression");
+  params.addClassDescription(
+      "Uses a parsed expression to set an extra element id for elements (via their centroids).");
+
+  return params;
+}
+
+ParsedExtraElementIDGenerator::ParsedExtraElementIDGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    FunctionParserUtils<false>(parameters),
+    _input(getMesh("input")),
+    _function(parameters.get<std::string>("expression")),
+    _eeid_names(getParam<std::vector<ExtraElementIDName>>("extra_element_id_names")),
+    _elem_id_name(getParam<std::string>("extra_elem_integer_name"))
+{
+  // base function object
+  _func_F = std::make_shared<SymFunction>();
+
+  // set FParser internal feature flags
+  setParserFeatureFlags(_func_F);
+
+  // add the constant expressions
+  addFParserConstants(_func_F,
+                      getParam<std::vector<std::string>>("constant_names"),
+                      getParam<std::vector<std::string>>("constant_expressions"));
+
+  // add the extra element integers
+  std::string symbol_str = "x,y,z";
+  for (const auto & eeid_name : _eeid_names)
+    symbol_str += "," + eeid_name;
+
+  // parse function
+  if (_func_F->Parse(_function, symbol_str) >= 0)
+    mooseError("Invalid function\n",
+               _function,
+               "\nin ParsedExtraElementIDGenerator ",
+               name(),
+               ".\n",
+               _func_F->ErrorMsg());
+
+  _func_params.resize(3 + _eeid_names.size());
+}
+
+std::unique_ptr<MeshBase>
+ParsedExtraElementIDGenerator::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  // the extra element ids would not have existed at construction so we only do this now
+  for (const auto & eeid_name : _eeid_names)
+    _eeid_indices.push_back(mesh->get_elem_integer_index(eeid_name));
+
+  std::set<SubdomainID> restricted_blocks;
+  if (isParamValid("restricted_subdomains"))
+  {
+    auto names = getParam<std::vector<SubdomainName>>("restricted_subdomains");
+    for (auto & name : names)
+    {
+      // check that the subdomain exists in the mesh
+      if (!MooseMeshUtils::hasSubdomainName(*mesh, name))
+        paramError("restricted_subdomains", "The block '", name, "' was not found in the mesh");
+
+      restricted_blocks.insert(MooseMeshUtils::getSubdomainID(name, *mesh));
+    }
+  }
+  else
+    mesh->subdomain_ids(restricted_blocks);
```

Comment:
this is not super efficient. I would just create a boolean and avoid the .count when there is no block restriction